api/client/commands.go

```diff
@@ -139,10 +140,38 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 		if _, err := os.Stat(root); err != nil {
 			return err
 		}
-		filename := path.Join(root, "Dockerfile")
+
+		absRoot, err := filepath.Abs(root)
+		if err != nil {
+			return err
+		}
+
+		var filename string
+		origDockerfile := *dockerfileName
+
+		if cmd.IsSet("f") || cmd.IsSet("file") {
+			if filename, err = filepath.Abs(*dockerfileName); err != nil {
+				return err
+			}
+		} else {
+			filename = path.Join(absRoot, *dockerfileName)
+		}
```

Comment:
Actually, thinking more about how the flag stuff works, I think the user might need to see:
-f, --file=""   Name of Dockerfile (default is CONTEXT-ROOT/Dockerfile)